lib/spark.js

```diff
@@ -395,3 +443,65 @@ function mapErrorToStatusCode(err) {
   // Fallback code for unknown errors
   return 600
 }
+
+/**
+ * Assigns weights to providers based on their protocol and context ID and picks one at random.
+ * Providers with higher weights have a higher chance of being selected.
+ *
+ * Providers serving content using Bitswap protocol are filtered out.
+ *
+ * @param {Provider[]} providers
+ * @returns {Provider | undefined}
+ */
+export function pickRandomProvider(providers) {
+  const filteredProviders = providers.filter((provider) => provider.protocol !== 'bitswap')
+  if (!filteredProviders.length) return
+  const weightedProviders = weighProviders(filteredProviders)
+  return pickRandomWeightedItem(weightedProviders)
+}
+
+/**
+ * Assigns weights to providers based on their protocol and context ID.
+ *
+ * HTTP providers and those whose context ID starts with 'gHa' are given higher weights,
+ * hence having a higher chance of being selected.
+ *
+ * @param {Provider[]} providers
+ * @returns {Provider & { weight: number }[]}
+ */
+function weighProviders(providers) {
+  const protocolWeights = { http: 2, graphsync: 1 }
+
+  // assign weight to each provider
+  return providers.map((provider) => {
+    let weight = protocolWeights[provider.protocol]
+    if (provider.contextId.startsWith('ghsA')) weight += 1
```

Comment:
I find it a bit difficult to reason about this weighting function, as I have to interpret this code to understand which combination get which weight.

IIUC, `protocol:graphsync & contextId:ghsA` and `protocol:http & contextId:(not ghsA)` have the same weights. Is that intentional?

I find it very cryptic why we are checking whether contextId starts with `ghsA`. Please add a comment to explain that such contextId values are likely to be created from the Filecoin PieceSize + PieceCID, therefore the retrieval provider is likely a Filecoin SP.